src/ir/possible-contents.cpp

```diff
@@ -0,0 +1,1607 @@
+      for (auto& elementSegment : wasm.elementSegments) {
+        for (auto* curr : elementSegment->data) {
+          if (auto* refFunc = curr->dynCast<RefFunc>()) {
+            calledFromOutside(refFunc->func);
+          }
+        }
+      }
+    }
+  }
+
+#ifdef POSSIBLE_CONTENTS_DEBUG
+  std::cout << "func phase\n";
+#endif
+
+  // Connect function parameters to their signature, so that any indirect call
+  // of that signature will reach them.
+  // TODO: find which functions are even taken by reference
+  for (auto& func : wasm.functions) {
+    for (Index i = 0; i < func->getParams().size(); i++) {
+      links.insert(getIndexes({SignatureParamLocation{func->type, i},
+                               LocalLocation{func.get(), i, 0}}));
+    }
+    for (Index i = 0; i < func->getResults().size(); i++) {
+      links.insert(getIndexes({ResultLocation{func.get(), i},
+                               SignatureResultLocation{func->type, i}}));
+    }
+  }
+
+#ifdef POSSIBLE_CONTENTS_DEBUG
+  std::cout << "struct phase\n";
+#endif
+
+  if (getTypeSystem() == TypeSystem::Nominal) {
+    subTypes = std::make_unique<SubTypes>(wasm);
+  }
+
+#ifdef POSSIBLE_CONTENTS_DEBUG
+  std::cout << "Link-targets phase\n";
+#endif
+
+  // Add all links to the targets vectors of the source locations, which we will
+  // use during the flow.
+  for (auto& link : links) {
+    getTargets(link.from).push_back(link.to);
+  }
+
+#ifndef NDEBUG
+  // Each vector of targets (which is a vector for efficiency) must have no
+  // duplicates.
+  for (auto& info : locations) {
+    disallowDuplicates(info.targets);
+  }
+#endif
+
+#ifdef POSSIBLE_CONTENTS_DEBUG
+  std::cout << "roots phase\n";
+#endif
+
+  // Set up the roots, which are the starting state for the flow analysis: send
+  // their initial content to them to start the flow.
+  for (const auto& [location, value] : roots) {
+#if defined(POSSIBLE_CONTENTS_DEBUG) && POSSIBLE_CONTENTS_DEBUG >= 2
+    std::cout << "  init root\n";
+    dump(location);
+    value.dump(std::cout, &wasm);
+    std::cout << '\n';
+#endif
+
+    updateContents(location, value);
+  }
+
+#ifdef POSSIBLE_CONTENTS_DEBUG
+  std::cout << "flow phase\n";
+  size_t iters = 0;
+#endif
+
+  // Flow the data while there is still stuff flowing.
+  while (!workQueue.empty()) {
+#ifdef POSSIBLE_CONTENTS_DEBUG
+    iters++;
+    if ((iters & 255) == 0) {
+      std::cout << iters++ << " iters, work left: " << workQueue.size() << '\n';
+    }
+#endif
+
+    auto iter = workQueue.begin();
+    auto locationIndex = *iter;
+    workQueue.erase(iter);
+
+    flowAfterUpdate(locationIndex);
+
+    updateNewLinks();
+  }
+
+  // TODO: Add analysis and retrieval logic for fields of immutable globals,
+  //       including multiple levels of depth (necessary for itables in j2wasm).
+}
+
+bool Flower::updateContents(LocationIndex locationIndex,
+                            PossibleContents newContents) {
```

Comment:
There is one place where we can actually remove targets and get a benefit from it - where we send stuff to the targets in the main loop. All the other places are different in some way, like when we process roots we will never visit them again, so there is nothing to remove. And when we compute the dynamic targets for StructGet etc., because the list is dynamic there is no benefit to pruning it (we throw that dynamic list away immediately).